tests/unit/test_charm.py

```diff
@@ -40,7 +42,7 @@ def setUp(self):
         self.harness.begin()
         # we need to have this to sync up the charm state: i.e. the
         # _stored.config
-        with patch("builtins.open", new_callable=mock_open):
+        with patch("charm.open", new_callable=mock_open):
```

Comment:
Just a comment (non-blocking): this patch `charm.open` is used almost in all tests case, do you think it maybe better to place it in `setUp` or `setUpClass`?

Reply:
Yes, maybe. I do not want to change more thing that are needed, however since I switch to use jinja2 env it breaks all test if we mocked open for all packages.
However to be honest, all these test are really messy and I think all should be reworked, if we want to maintain this charm.

Reply:
That's totally understandable